server/src/main/java/io/crate/replication/logical/MetadataTracker.java

```diff
@@ -207,48 +205,46 @@ private CompletableFuture<?> processSubscription(String subscriptionName) {
         } catch (Exception e) {
             return CompletableFuture.failedFuture(e);
         }
-        return getRemoteClusterState(client)
-            .thenCompose(remoteState -> {
-                PublicationsMetadata publicationsMetadata = remoteState.metadata().custom(PublicationsMetadata.TYPE);
-                if (publicationsMetadata == null) {
-                    LOGGER.trace("No publications found on remote cluster.");
-                    return CompletableFuture.completedFuture(false);
-                }
-                return updateClusterState(subscriptionName, subscription, remoteState);
-            }).thenCompose(acked -> {
-                if (!acked) {
-                    return CompletableFuture.completedFuture(false);
-                }
-                var request = new PublicationsStateAction.Request(
-                    subscription.publications(),
-                    subscription.connectionInfo().user()
-                );
-                return client.execute(PublicationsStateAction.INSTANCE, request)
-                    .thenCompose(publicationsStateResp -> subscribeToNewRelations(
-                        subscriptionName,
-                        subscription,
-                        subscriberState,
-                        publicationsStateResp,
-                        replicationService));
-            }).exceptionallyCompose(err -> {
-                var e = SQLExceptions.unwrap(err);
-                if (SQLExceptions.maybeTemporary(e)) {
-                    LOGGER.warn("Retrieving remote metadata failed for subscription '" + subscriptionName + "', will retry", e);
-                    return CompletableFuture.completedFuture(null);
-                }
-                var msg = "Tracking of metadata failed for subscription '" + subscriptionName + "'" + " with unrecoverable error, stop tracking";
-                LOGGER.error(msg, e);
-                return replicationService.updateSubscriptionState(subscriptionName, Subscription.State.FAILED, msg + ".\nReason: " + e.getMessage())
-                    .handle((ignoredAck, ignoredErr) -> {
-                        stopTracking(subscriptionName);
-                        return null;
-                    });
-            });
+
+        var request = new PublicationsStateAction.Request(
+            subscription.publications(),
+            subscription.connectionInfo().user()
+        );
+        CompletableFuture<Response> publicationsState = client.execute(PublicationsStateAction.INSTANCE, request);
+        CompletableFuture<Boolean> updatedClusterState = publicationsState.thenCompose(response ->
+            updateClusterState(subscriptionName, subscription, response)
+        );
+        return updatedClusterState.thenCompose(acked -> {
+            if (!acked) {
+                return CompletableFuture.completedFuture(false);
+            }
+            var publicationResponse = publicationsState.join();
```

Comment:
It's not blocking, if `updatedClusterState.thenCompose` triggers then it is implied that `publicationsState` already completed - otherwise `updatedClusterState` wouldn't have triggered the `thenCompose` function.

The intention here was just to avoid having to pass through the publicationStateResponse through the earlier thenCompose chain which would have record a tuple/record to have access to both, the response and the clusterState-acked value.

Reply:
I added an assertion that should help clarify it.